File: scripts/blocks/effect/elementAura.js
const fireAuraRange = 15 * Vars.tilesize;

/* Fire Aura effect */
const fireAuraColor = "ffaa44";
const fireAuraEffect = newEffect(40, e => {
  Draw.color(Color.valueOf(fireAuraColor), Pal.darkFlame, e.fin());
  Angles.randLenVectors(e.id, 3, 2 + e.fin() * 9, cons((x, y) => {
    Fill.circle(e.x + x, e.y + y, 0.2 + e.fslope() * 1.5);
  }));
});

/* Aura bullet */
const fireAuraBullet = extend(BulletType, {});
fireAuraBullet.bulletSprite = Core.atlas.find("clear");
fireAuraBullet.speed = 0.001;
fireAuraBullet.lifetime = 1;
fireAuraBullet.damage = 0;
fireAuraBullet.status = StatusEffects.melting;

/* Fire Aura */
const fireAura = extendContent(PowerTurret, "fire-aura", {
  load(){
    this.super$load();
    this.liquidRegion = Core.atlas.find(this.name + "-liquid");
    this.topRegion = Core.atlas.find(this.name + "-top");
  },
  update(tile){
    var entity = tile.ent();
    if(!this.validateTarget(tile)){
      entity.target = null;
    }
    entity.recoil = 0;
    entity.rotation = 0;
    entity.heat = Mathf.lerpDelta(entity.heat,
      this.validateTarget(tile) ? 1 : 0,
      this.validateTarget(tile) ? this.warmup : this.cooldown
    );
    if (entity.cons.valid()){
      if(entity.timer.get(this.timerTarget, this.targetInterval)){
        this.findTarget(tile);
      }
      if(this.validateTarget(tile)){
        this.updateShooting(tile);
      }
    }
  },
  drawLayer(tile){
    this.super$drawLayer(tile);
    Draw.color(entity.liquids.current().color);
    Draw.alpha(entity.liquids.total() / this.liquidCapacity);
    Draw.rect(this.liquidRegion, tile.drawx(), tile.drawy());
    Draw.color();
    Draw.rect(this.topRegion, tile.drawx(), tile.drawy());
  },
  updateShooting(tile){
    var entity = tile.ent();
    if(entity.reload >= this.reload){
      var type = this.peekAmmo(tile);
      this.shoot(tile, type);
      entity.reload = 0;
    }else{
      entity.reload += tile.entity.delta() * (this.peekAmmo(tile)).reloadMultiplier * this.baseReloadSpeed(tile);
    }
  },
  shoot(tile, type){
    var entity = tile.ent();
    var radius = this.range;
    Units.nearbyEnemies(tile.getTeam(), tile.drawx() - radius, tile.drawy() - radius, radius * 2, radius * 2, cons(unit => {
      if(unit.withinDst(tile.drawx(), tile.drawy(), radius)){
        if(!unit.isDead() && unit instanceof HealthTrait){
          Calls.createBullet(type, tile.geatTeam(), unit.getX(), unit.getY(), 0, 1, 1);
        }
      }
    }));
    this.effects(tile);
    this.effectsArea(tile, this.effectAreaCount);
    this.useAmmo(tile);
  },
  effects(tile){
    var shootEffect = this.shootEffect == Fx.none ? (this.peekAmmo(tile)).shootEffect : this.shootEffect;
    var smokeEffect = this.smokeEffect == Fx.none ? (this.peekAmmo(tile)).smokeEffect : this.smokeEffect;
    var entity = tile.ent();
    Effects.effect(shootEffect, tile.drawx(), tile.drawy(), entity.rotation);
    Effects.effect(smokeEffect, tile.drawx(), tile.drawy(), entity.rotation);
    this.shootSound.at(tile, Mathf.random(0.9, 1.1));
  },
  effectsArea: function(tile, i){
    var shootEffect = this.shootEffect == Fx.none ? (this.peekAmmo(tile)).shootEffect : this.shootEffect;
    var entity = tile.ent();
    for (var j = 0; j < i; j++){
      Effects.effect(shootEffect,
        tile.drawx() + Angles.trnsx(Mathf.random(360), Mathf.random(this.range)),
        tile.drawy() + Angles.trnsy(Mathf.random(360), Mathf.random(this.range)),
        entity.rotation
      );
    }
  },
  useAmmo: function(tile){
    tile.entity.cons.trigger();
    return this.super$useAmmo(tile);
  },
  hasAmmo: function(tile){
    return tile.entity.cons.valid();
  },
  shouldTurn: function(tile){
    return false;
  },
});
/*const fireAura = extendContent(Turret, "fire-aura", {
  load(){
    this.super$load();
    this.liquidRegion = Core.atlas.find(this.name + "-liquid");
    this.topRegion = Core.atlas.find(this.name + "-top");
  },
  update(tile){
    var entity = tile.ent();
    if (!this.validateTarget(tile)){
      entity.target = null;
    }
    entity.recoil = 0;
    entity.heat = Mathf.lerpDelta(entity.heat,
      this.validateTarget(tile) ? 1 : 0,
      this.validateTarget(tile) ? this.warmup : this.cooldown
    );
    if (this.hasAmmo(tile)){
      if (entity.timer.get(this.timerTarget, this.targetInterval)){
        this.findTarget(tile);
      }
      if (this.validateTarget(tile)){
        if (entity.rotation == null){
          entity.rotation = 0;
        }
        this.updateShooting(tile);
      }
    }
  },
  drawLayer(tile){
    this.super$drawLayer(tile);
    Draw.color(entity.liquids.current().color);
    Draw.alpha(entity.liquids.total() / this.liquidCapacity);
    Draw.rect(this.liquidRegion, tile.drawx(), tile.drawy());
    Draw.color();
    Draw.rect(this.topRegion, tile.drawx(), tile.drawy());
  },
  shoot(tile, type){
    Calls.createBullet(type, tile.getTeam(), tile.drawx(), tile.drawy(), Mathf.random(360), 1, 1);
    this.effects(tile);
    this.effectsArea(tile, this.effectAreaCount);
    this.useAmmo(tile);
  },
  useAmmo: function(tile){
    var entity = tile.ent();
    if (tile.isEnemyCheat()){
      return this.shootType;
    }
    var type = this.shootType;
    entity.cons.trigger();
    return type;
  },
  peekAmmo: function(tile){
    return this.shootType;
  },
  hasAmmo: function(tile){
    var entity = tile.ent();
    return entity.cons.valid();
  },
  setStats(){
    this.super$setStats();
    this.stats.add(BlockStat.damage, this.shootType.damage, StatUnit.none);
  },
  shouldTurn: function(tile){
    return false;
  },
  baseReloadSpeed(tile){
    return tile.isEnemyCheat() ? 1 : tile.entity.power.status;
  },
});*/
fireAura.reload = 5;
fireAura.shootType = fireAuraBullet;
fireAura.range = fireAuraRange;
fireAura.effectAreaCount = 3;
fireAura.hasLiquids = true;
fireAura.liquidCapacity = 20;
fireAura.shootEffect = fireAuraEffect;
fireAura.smokeEffect = Fx.fireSmoke;
fireAura.ammoUseEffect = Fx.none;
fireAura.targetInterval = 5;
fireAura.warmup = 0.08;

File: src/mw/world/blocks/defense/ElementAura.java
package mw.world.blocks.defense;

import arc.*;
import arc.graphics.g2d.*;
import arc.math.*;
import mindustry.content.*;
import mindustry.entities.*;
import mindustry.entities.bullet.*;
import mindustry.gen.*;
import mindustry.type.*;
import mindustry.world.*;
import mindustry.world.consumers.*;
import mindustry.world.meta.*;
import mindustry.world.meta.values.*;

import static mindustry.Vars.*;

public class ElementAura extends Block{
    protected BulletType bullet;

    public Liquid liquid;
    public TextureRegion liquidRegion;

    public float warmup = 0.08f;
    public float cooldown = 0.02f;
    public float liquidUsed = 1f;
    public float reloadTime = 5f;

    public int areaEffectCount = 3;
    public Effect shootEffect = Fx.none;
    public Effect smokeEffect = Fx.none;

    public final int timerTarget = timers++;
    public float targetInterval = 5f;

    public float range = 15f * tilesize;

    public ElementAura(String name, Liquid liquid){
        super(name);
        this.liquid = liquid;
        outlineIcon = false;
    }

    public ElementAura(String name){
        this(name, Liquids.water);
    }

    @Override
    public void setStats(){
        super.setStats();

        if(consumes.has(ConsumeType.liquid)){
            consumes.remove(ConsumeType.liquid);
        }
        stats.add(BlockStat.input, new LiquidValue(liquid, 60 / reloadTime, true));
    }

    @Override
    public void load(){
        super.load();

        liquidRegion = Core.atlas.find(name + "-liquid");
    }

    public class ElementAuraBuild extends Building{
        public float heat = 0.0f;
        public float reload = reloadTime;

        public Teamc target;
        public boolean hasTarget = false;

        protected boolean hasShot = false;

        @Override
        public void updateTile(){
            if(!validateTarget()){
                target = null;
                hasTarget = false;
            }

            heat = Mathf.lerpDelta(heat, 
                shooting() ? 1 : 0,
                shooting() ? warmup : cooldown
            );

            if(hasAmmo()){
                if(timer(timerTarget, targetInterval)){
                    findTarget();
                }

                if(shooting()){
                    updateShooting();
                }
            }
        }

        public void updateShooting(){
            if(reload <= 0.0f){
                shoot();
                reload = reloadTime;
            }else{
                reload = Math.max(reload - edelta() * timeScale(), 0f);
            }
        }

        public boolean shooting(){
            return hasAmmo() && hasTarget;
        }

        public boolean validateTarget(){
            return !Units.invalidateTarget(target, team, x, y, range);
        }

        public boolean hasAmmo(){
            return liquids.get(liquid) >= liquidUsed * timeScale();
        }

        public void findTarget(){
            target = Units.closestTarget(team(), x, y, range, e -> !e.dead());
            hasTarget = target != null;
        }

        public void shoot(){
            hasShot = false;

            Units.nearbyEnemies(team(), x - range, y - range, range * 2, range * 2, unit -> {
                if(unit.within(this, range) && !unit.dead()){
                    bullet.create(this, team(), unit.x, unit.y, 0f, 0f, 0f);

                    if(!hasShot){
                        effects();
                        useAmmo();

                        hasShot = true;
                    }
                }
            });
        }

        public void effects(){
            Effect shoot = shootEffect == Fx.none ? bullet.shootEffect : shootEffect;
            Effect smoke = smokeEffect == Fx.none ? bullet.smokeEffect : smokeEffect;

            for(int i = 0; i < areaEffectCount; i++){
                float x = this.x + Angles.trnsx(Mathf.random(360f), Mathf.random(range));
                float y = this.y + Angles.trnsy(Mathf.random(360f), Mathf.random(range));

                shoot.at(x, y);
                smoke.at(x, y);
            }
        }

        public BulletType useAmmo(){
            if(cheating()) return bullet;

            liquids.remove(liquid, liquidUsed);

            return bullet;
        }

        @Override
        public boolean acceptItem(Building source, Item type){
            return false;
        }

        @Override
        public boolean acceptLiquid(Building source, Liquid type, float amount) {
            return liquid == type
                && liquids.get(liquids.current()) <= liquidUsed;
        }
    }
}
